app/javascript/components/Admin/Purchases/index.tsx

```diff
@@ -0,0 +1,697 @@
+          url={Routes.refund_admin_cards_path({ stripe_fingerprint: purchase.stripe_fingerprint })}
+          loading="Refunding..."
+          done="Refunding purchases!"
+          confirm_message="Are you sure you want to Mass-refund for fraud all purchases associated with this purchase's card?"
+          success_message="Refunding purchases!"
+        />
+      </>
+    ) : null}
+    {purchase.subscription &&
+    !purchase.subscription.cancelled &&
+    !purchase.subscription.ended_at &&
+    !purchase.subscription.failed_at ? (
+      <>
+        <AdminActionButton
+          label="Cancel subscription for buyer"
+          url={Routes.cancel_subscription_admin_purchase_path(purchase, { by_seller: false })}
+          loading="Canceling..."
+          done="Canceled!"
+          confirm_message="Are you sure you want to cancel this subscription on behalf of the buyer?"
+          success_message="Canceled!"
+        />
+        <AdminActionButton
+          label="Cancel subscription for seller"
+          url={Routes.cancel_subscription_admin_purchase_path(purchase, { by_seller: true })}
+          loading="Canceling..."
+          done="Canceled!"
+          confirm_message="Are you sure you want to cancel this subscription on behalf of the seller?"
+          success_message="Canceled!"
+        />
+      </>
+    ) : null}
+    {purchase.buyer_blocked ? (
+      <AdminActionButton
+        label="Unblock buyer"
+        url={Routes.unblock_buyer_admin_purchase_path(purchase)}
+        loading="Unblocking buyer..."
+        done="Buyer unblocked!"
+        success_message="Buyer unblocked!"
+      />
+    ) : (
+      <AdminActionButton
+        label="Block buyer"
+        url={Routes.block_buyer_admin_purchase_path(purchase)}
+        loading="Blocking buyer..."
+        done="Buyer blocked!"
+        confirm_message="This will fully block this buyer's emails, GUID, and IP addresses. Proceed?"
+        success_message="Buyer blocked!"
+      />
+    )}
+    {purchase.is_deleted_by_buyer ? (
+      <AdminActionButton
+        label="Undelete"
+        url={Routes.undelete_admin_purchase_path(purchase)}
+        loading="Undeleting..."
+        done="Undeleted!"
+        confirm_message="Are you sure you want to undelete this purchase?"
+        success_message="Undeleted!"
+      />
+    ) : null}
+    {purchase.successful ? (
+      <Link href={Routes.receipt_purchase_path(purchase.external_id)} target="_blank" className="button small">
+        Go to Receipt
+      </Link>
+    ) : null}
+  </div>
+);
+
+const PurchaseUrlRedirect = ({
+  url_redirect,
+  label = "URL redirect",
+}: {
+  url_redirect: UrlRedirect;
+  label?: string;
+}) => (
+  <>
+    <dt>{label}</dt>
+    <dd>
+      <Link href={url_redirect.download_page_url} target="_blank">
+        {url_redirect.download_page_url}
+      </Link>{" "}
+      ({url_redirect.uses} uses)
+    </dd>
+  </>
+);
+
+const AdminPurchase = ({ purchase }: { purchase: Purchase }) => (
```

Comment:
Do you have a (strong?) opinion on whether the "main" component should be at the top or bottom? Personally I have a small preference towards bottom so declaration is before use, but either would work.